### PR TITLE
feat: added view config_type_summary

### PR DESCRIPTION
**Pull request by adityathebe**
Resolves https://github.com/flanksource/duty/issues/122

### CODE REVIEW

**Comment by adityathebe** [comment]
```
+-------------------+--------------------------------------------------------+---------+---------------+-----------------+---------------+---------------+----------------+
| config_type       | analysis                                               | changes | total_configs | cost_per_minute | cost_total_1d | cost_total_7d | cost_total_30d |
|-------------------+--------------------------------------------------------+---------+---------------+-----------------+---------------+---------------+----------------|
|                   | <null>                                                 | 1       | 1             | <null>          | <null>        | <null>        | <null>         |
| Account           | {"security": 1}                                        | <null>  | 1             | <null>          | <null>        | <null>        | <null>         |
| Config            | <null>                                                 | <null>  | 3             | <null>          | <null>        | <null>        | <null>         |
| ContainerRegistry | <null>                                                 | <null>  | 2             | <null>          | <null>        | <null>        | <null>         |
| DHCP              | <null>                                                 | <null>  | 5             | <null>          | <null>        | <null>        | <null>         |
| DiskStorage       | {"cost": 15, "reliability": 23}                        | <null>  | 25            | <null>          | <null>        | <null>        | <null>         |
| DNSZone           | {"security": 1, "performance": 2, "reliability": 3}    | <null>  | 7             | <null>          | <null>        | <null>        | <null>         |
| KubernetesCluster | {"security": 1, "performance": 2, "recommendation": 2} | <null>  | 3             | <null>          | <null>        | <null>        | <null>         |
| LoadBalancer      | <null>                                                 | <null>  | 3             | <null>          | <null>        | <null>        | <null>         |
| ObjectStorage     | <null>                                                 | <null>  | 21            | <null>          | <null>        | <null>        | <null>         |
| Profile           | <null>                                                 | <null>  | 4             | <null>          | <null>        | <null>        | <null>         |
| Region            | {"cost": 3, "security": 19, "reliability": 5}          | <null>  | 18            | <null>          | <null>        | <null>        | <null>         |
| ResourceGroup     | <null>                                                 | <null>  | 10            | <null>          | <null>        | <null>        | <null>         |
| Role              | <null>                                                 | <null>  | 49            | <null>          | <null>        | <null>        | <null>         |
| Route             | <null>                                                 | <null>  | 8             | <null>          | <null>        | <null>        | <null>         |
| SecurityGroup     | {"security": 19}                                       | 25      | 23            | <null>          | <null>        | <null>        | <null>         |
| Subnet            | <null>                                                 | 17      | 24            | <null>          | <null>        | <null>        | <null>         |
| User              | {"security": 11}                                       | <null>  | 16            | <null>          | <null>        | <null>        | <null>         |
| VirtualMachine    | {"cost": 2, "security": 2}                             | 55      | 5             | <null>          | <null>        | <null>        | <null>         |
| VirtualNetwork    | {"availability": 3}                                    | <null>  | 3             | <null>          | <null>        | <null>        | <null>         |
| VPC               | {"security": 2}                                        | 16      | 6             | <null>          | <null>        | <null>        | <null>         |
+-------------------+--------------------------------------------------------+---------+---------------+-----------------+---------------+---------------+----------------+
```

File: fixtures/dummy/config.go [hunk 1]
```diff
@@ -6,43 +6,55 @@ import (
...
-	ID: uuid.New(),
+	ID:           uuid.New(),
+	ConfigType:   "kubernetesNode",
+	CostTotal30d: 1,
 }
 
 var KubernetesNodeB = models.ConfigItem{
-	ID: uuid.New(),
+	ID:           uuid.New(),
+	ConfigType:   "kubernetesNode",
```
Comment (yashmehrotra) [inline fixtures/dummy/config.go:26]: Make these Capitalized Constants in models/config_item.go

We will eventually use them directly in conifg-db

File: config_type_summary_test.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+				Expect(err).ToNot(HaveOccurred())
+			}
+
+			configTypeSummaries = append(configTypeSummaries, c)
+		}
+
+		Expect(configTypeSummaries).To(HaveLen(5))
+		Expect(configTypeSummaries).To(Equal([]configTypeSummary{
+			{
+				configType:   "database",
```
Comment (yashmehrotra) [inline config_type_summary_test.go:43]: Use the same constants here

**Review by yashmehrotra — CHANGES_REQUESTED** [review]

**Review by yashmehrotra — CHANGES_REQUESTED** [review]
LGTM, just update the constants

File: models/config_item.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package models
+
+// Config Types
+const (
```
Comment (moshloop) [inline models/config_item.go:4]: I don't think these shold be global - maybe just have them in the dummy pacakge - Also don't use acronoym i.e. ConfigTypeNode

File: views/006_config_views.sql [hunk 4]
```diff
@@ -180,3 +180,57 @@ DROP VIEW IF EXISTS config_tags;
...
+DROP VIEW IF EXISTS config_type_summary;
+CREATE VIEW config_type_summary AS
+  WITH changes_per_type AS (
+    SELECT
+      config_items.config_type,
+      COUNT(config_changes.id) AS count
+    FROM
+      config_changes
+      LEFT JOIN config_items ON config_changes.config_id = config_items.id
+    GROUP BY
```
Comment (moshloop) [inline views/006_config_views.sql:194]: Can we limit these to changes created in last 30 days ?

Comment (adityathebe) [reply views/006_config_views.sql:195]: made the change

**Comment by yashmehrotra** [comment]
@moshloop Lets keep the `ConfigTypeNode`, `ConfigTypeCluster` etc. in `models/config_items.go`, that way we can directly use them in config-db's scraper as const types

**Comment by moshloop** [comment]
>  Lets keep the `ConfigTypeNode`, `ConfigTypeCluster` etc. in `models/config_items.go`, that way we can directly use them in config-db's scraper as const types as `models.ConfigTypeNode`

@yashmehrotra - the problem is that these are not really constant, and maintaining that list here implies something

There is perhaps a concept of "common" type we can maintain a list of e.g. VirtualMachine, KubernetesCluster, NetworkRule and then in the config scraper it would be:

```go
ConfigType: models.ConfigTypeVirtualMachine,
ExternalConfigType: "AWS::EC2"
```


Maybe we should rename config type to config class and externalconfigtype to configtype ?

**Comment by yashmehrotra** [comment]
> @yashmehrotra - the problem is that these are not really constant, and maintaining that list here implies something
> There is perhaps a concept of "common" type we can maintain a list of e.g. VirtualMachine, KubernetesCluster, NetworkRule and then in the config s craper it would be:
> ```
> ConfigType: models.ConfigTypeVirtualMachine,
> ExternalConfigType: "AWS::EC2"
> ```

Exactly, the source of truth for the types (VirtualMachine, Cluster, Node,LoadBalancer) should reside in duty

>Maybe we should rename config type to config class and externalconfigtype to configtype ?
Config class sounds good, it is explicit and makes sense

**Comment by yashmehrotra** [comment]
> > Exactly, the source of truth for the types (VirtualMachine, Cluster, Node,LoadBalancer) should reside in duty
> 
> Except that we have a ConfigTypeEC2 - If we are going to rename the column to config class then it probably makes sense todo it now.

Yeah, BTW, in config-db, I had changed EC2 type to VirtualMachine and other AWS Types to generic when I was working on Azure

image

**Review by yashmehrotra — APPROVED** [review]

**Pull request merged**